### PR TITLE
Add commit comment when push to pull request branch

### LINKED ISSUE #2830: Show commits to a branch in PR
When a user pushes a commit to a branch which has an open PR, this commit should appear as a comment in the PR. Basically like it does in Github.

This is probably "nice to have" but after working with Github for some time, it bothers me somehow that this isn't implemented in Gitea.

(I don't know if there already is an issue about this, if yes please point me to it.)

### PR DESCRIPTION
**Pull request by lunny**
Fixes #2830

### CODE REVIEW

**Comment by ethantkoenig** [comment]
Should these comments be removed on a force push (like in Github)?

**Comment by lunny** [comment]
@ethantkoenig yes, I think you are right. I will update this.

File: models/issue_comment.go
Comment (jonasfranz) [inline models/issue_comment.go]: Do we need a migration for this?

Comment (lunny) [reply models/issue_comment.go]: What's migration? We didn't change anything, the extra fields only for new comment type.

File: models/action.go
Comment (lafriks) [inline models/action.go]: Do we really need to store link to commit? I think we should generate it on the fly as it is internal url and will become invalid if user changes Gitea configuration

**Review by lafriks — CHANGES_REQUESTED** [review]

File: models/issue_comment.go
Comment (jonasfranz) [reply models/issue_comment.go]: You added the Link column. Do we need a migration for syncing this?

File: models/action.go
Comment (lunny) [reply models/action.go]: If we didn't save link before, then we have to read repo table to know the repo path for every comment. I don't think that's necessary.

Comment (lunny) [reply models/action.go]: The link is a relative link not an absolute link.

File: models/issue_comment.go
Comment (lunny) [reply models/issue_comment.go]: Currently Link is only for the added comment type.

File: models/action.go
Comment (lafriks) [reply models/action.go]: yes but it will include AppSubURL that can be changed in settings. Maybe save only issue.Repo.FullName() in link and create new method for comment type that depending on comment type would generate URL (in this case AppSubURL + "/" + comment.Link + "/commit/" + comment.Sha1).
Additional change needed is to update link column on user/org rename and repository rename

**Review by lafriks — CHANGES_REQUESTED** [review]

File: models/action.go
Comment (lunny) [reply models/action.go]: done.

**Comment by lunny** [comment]
@ethantkoenig done.

File: templates/repo/issue/view_content/comments.tmpl
Comment (lafriks) [inline templates/repo/issue/view_content/comments.tmpl]: Please create method for Comment type to generate target URL and use it here and lower

**Review by lafriks — CHANGES_REQUESTED** [review]

File: models/action.go
Comment (daviian) [inline models/action.go]: Can be simplified to `if !isNewBranch && git.EmptySHA != opts.Commit.Commits[0].Sha1 {`

File: routers/repo/issue.go
Comment (daviian) [inline routers/repo/issue.go]: Can we replace this if construct by a switch statement?

File: templates/repo/issue/view_content/comments.tmpl
Comment (daviian) [inline templates/repo/issue/view_content/comments.tmpl]: Can we keep consistency either using camel case or upper case?

**Review by daviian — CHANGES_REQUESTED** [review]

File: models/issue_comment.go
Comment (daviian) [inline models/issue_comment.go]: Why does this PR change code regarding `CommitStatus`?

File: models/pull.go
Comment (daviian) [inline models/pull.go:69]: Why do we need `LastCommitID`? It is only used once in assignment.

**Review by daviian — CHANGES_REQUESTED** [review]

File: models/pull.go
Comment (lunny) [reply models/pull.go:69]: @daviian #2519 will need it.

Comment (lunny) [reply models/pull.go:69]: It will help to show the pull's commit status. If we didn't cache it, we will spent many time to invoke `git` command on that PR.

Comment (daviian) [reply models/pull.go:69]: Ok thanks for clarification 👍

**Comment by lunny** [comment]
@daviian @lafriks done and rebased.

**Review by lafriks — APPROVED** [review]

**Review by daviian — APPROVED** [review]

**Comment by lafriks** [comment]
BTW what about updating full name when user/org is renamed or repository is renamed/transfered/deleted etc?

File: models/action.go
Comment (ethantkoenig) [inline models/action.go]: I don't think this is right. Suppose I push 3 commits on top of a branch's current head (not a force push)
```
sha1 newest (my parent is sha2)
sha2 older (my parent is sha3)
sha3 even older (my parent is the previous head of the branch)
```

`opts.Commits.Commits` will contain `sha1`, `sha2` and `sha3`, in that order. We will run `git rev-list sha1 ^sha3`, which will return `sha1` and `sha2`, and this push will incorrectly be marked as a force push.

File: models/action.go
Comment (ethantkoenig) [inline models/action.go]: nit: `for _, c := range opts.Commits.Commits`

File: models/issue_comment.go
Comment (ethantkoenig) [inline models/issue_comment.go]: I don't think this right. Suppose there are two commits that have previously been pushed to a PR:
```
sha1 newest (my parent is sha2)
sha2 older (my parent is something else)
```
Suppose `sha1` is removed via a force push, but `sha2` remains. `sha2`'s comment should not be deleted. Even if we recreate it later, the new comment will have a different timestamp than the old deleted one, which affects the in order in which comments are listed.

File: templates/repo/commit_status.tmpl
Comment (ethantkoenig) [inline templates/repo/commit_status.tmpl:1]: This file uses spaces

**Review by ethantkoenig — CHANGES_REQUESTED** [review]

File: models/action.go
Comment (daviian) [reply models/action.go]: I think it should be this:
`output, err := git.NewCommand("rev-list", opts.OldCommitID, "^"+opts.NewCommitID).RunInDir(repo.RepoPath())`

**Comment by lafriks** [comment]
@lunny can you resolve requested changes before 1.3?

**Comment by lunny** [comment]
@lafriks Yes, I will.

File: models/issue_comment.go
Comment (lunny) [reply models/issue_comment.go]: Since it's a force push the old commit maybe changed also. We cannot detect that. And it seems github also always delete all previous commit comments.

File: models/action.go
Comment (lunny) [reply models/action.go]: done.

File: models/action.go
Comment (lunny) [reply models/action.go]: done.

File: templates/repo/commit_status.tmpl
Comment (lunny) [reply templates/repo/commit_status.tmpl:1]: done.

**Comment by lunny** [comment]
@lafriks reabased.

File: models/action.go
Comment (thehowl) [inline models/action.go]: `var isForcePush bool`?

File: models/action.go
Comment (thehowl) [inline models/action.go]: `return` already interrupts the flow - so using the else if is unneccessary (remove else). Also, remove the if entirely and make it `isForcePush = len(output) > 0`?

File: models/action.go
Comment (thehowl) [inline models/action.go]: why no `:=`?

File: models/action.go
Comment (thehowl) [inline models/action.go]: from

File: models/issue_comment.go
Comment (thehowl) [inline models/issue_comment.go]: reverse it?

e.g.
```go
if c.CommitStatus != nil {
    return nil
}
```

Then everything else can be placed on just one line of indentation.

File: models/issue_comment.go
Comment (thehowl) [inline models/issue_comment.go]: remove the `else`? If code is reachable at this point, it always means `err == nil`, since the `err != nil` branch returns.

**Review by thehowl — CHANGES_REQUESTED** [review]
Mostly style nitpicks, I'm not quite comfortable with how things work on Gitea yet so I didn't really focus too much on trying to understand the logic.

**Comment by lunny** [comment]
@thehowl all done.

File: models/action.go
Comment (ethantkoenig) [inline models/action.go]: Typo: coment -> comment

File: models/issue_comment.go
Comment (ethantkoenig) [inline models/issue_comment.go]: If `repo` is already an argument, why does `repoFullName` need to be another argument?

Also, would it make more sense to have a `*Commit` argument instead of `content` and `commitSHA`?

File: models/issue_comment.go
Comment (ethantkoenig) [reply models/issue_comment.go]: We can detect it. The commits whose comments need to be deleted are exactly the commits listed in the output from the `git rev-list ...` command.

**Review by ethantkoenig — CHANGES_REQUESTED** [review]

File: models/issue_comment.go
Comment (ethantkoenig) [reply models/issue_comment.go]: We can detect it. The commits whose comments need to be deleted are exactly the commits listed in the output from the `git rev-list ...` command.

Also, from what I can tell Github does not delete unaffected commit comments.

File: models/action.go
Comment (lunny) [reply models/action.go]: done.

File: models/issue_comment.go
Comment (lunny) [reply models/issue_comment.go]: done.

**Comment by lafriks** [comment]
@lunny if forked repository or it's owner gets renamed `RepoFullName` needs to be updated to new url

**Comment by lunny** [comment]
@lafriks done and rebased.

**Comment by lunny** [comment]
@lafriks done.

**Comment by lafriks** [comment]
LGTM

**Comment by lafriks** [comment]
@ethantkoenig needs your approval

**Comment by ethantkoenig** [comment]
@lafriks @lunny One of my comments (https://github.com/go-gitea/gitea/pull/2548#discussion_r140647785) still needs to be addressed. I don't mean to block this PR, but I believe it will be relatively straightforward to fix.

File: models/issue_comment.go
Comment (lunny) [reply models/issue_comment.go]: @ethantkoenig could you give me some example pull request on github the commit comments are not deleted when force push?

Comment (ethantkoenig) [reply models/issue_comment.go]: @lunny
image

image

File: models/action.go
Comment (ethantkoenig) [inline models/action.go]: nit: remove this blank line

File: models/issue_comment.go
Comment (ethantkoenig) [inline models/issue_comment.go]: nit: remove this blank line

File: models/issue_comment.go
Comment (ethantkoenig) [inline models/issue_comment.go]: This field seems unnecessary; if we need to know what the repo's name is, we can find that out by using `c.Issue.Repo`. In fact, we already do this in a lot of places (e.g. [Comment.HTMLURL()](https://github.com/go-gitea/gitea/blob/240609432b84a8caa5ec4c99a916277952ef02d5/models/issue_comment.go#L153))

File: models/issue_comment.go
Comment (ethantkoenig) [inline models/issue_comment.go:638]: As I mentioned in [a previous comment](https://github.com/go-gitea/gitea/pull/2548#discussion_r147324379), we can determine which comments need to be deleted from the `git rev-list` command that is run in [CommitRepoAction(..)](https://github.com/go-gitea/gitea/pull/2548/files#diff-7603238519a500de37f3e4b25cef1800R645). I believe that would be simpler and more efficient (since we have to run `git rev-list` regardless) than using the database.

**Review by ethantkoenig — CHANGES_REQUESTED** [review]

File: models/issue_comment.go
Comment (lunny) [reply models/issue_comment.go:638]: I have tested some times. It seems `git rev-list` returned different commits from `opts.Commits.Commits`.

File: models/issue_comment.go
Comment (lunny) [reply models/issue_comment.go]: done.

File: models/action.go
Comment (lunny) [reply models/action.go]: done.

File: models/issue_comment.go
Comment (lunny) [reply models/issue_comment.go]: done.

File: models/issue_comment.go
Comment (ethantkoenig) [reply models/issue_comment.go:638]: I have also done several examples (e.g. just removing commits, removing commits and adding new ones), and in every case the commit shas returned by `git rev-list` were exactly the commits whose comments would need to be removed (i.e. the commits that previously existed but do not exist after the force push)

Do you have an example where `git rev-list` does not work?

Comment (lunny) [reply models/issue_comment.go:638]: Yes. `git rev-list` is easier than current method. But it may add the master commits to this PR. When I rebase and push to a feature test. Some of commits are already in master and another not. We only add commit comments that not in target branch. See below screenshots is what the method you recommend. `13fabd0` has been pushed to `master` branch. Then I rebase `master` to `test` and then push to `777:test` it has been pull requested. Then commit comment `13fabd0` occupied but it should not.

image

image

Comment (lafriks) [reply models/issue_comment.go:638]: @lunny I think he meant to use rev-list to detect commits that need to be deleted not added

Comment (lunny) [reply models/issue_comment.go:638]: Yes, I see. but it's not enough. For my example, `3fabd0261` should not be as a commit comment, because it has been submited to maser. rev-list will not remove `3fabd0261`. So if force push, we have to find the HEAD of target branch and remove all commits before that.

File: models/issue_comment.go
Comment (ethantkoenig) [inline models/issue_comment.go:676]: From what I can tell, this is not right.

Suppose I have the following (where later-in-the-alphabet commits have a later timestamp; it appears this affects the order of `opts.Commits`, which is itself kind of concerning)
```
A---B(master, origin/master)
 \
  C(develop, origin/develop)
```
and I rebase develop onto master, resulting in 
```
      C'(develop)
     /
A---B(master, origin/master)
 \
  C(origin/develop)
```

If I force push develop, the `opts.Commits` will contain `[C', C, B]`, in that order. In this case (since `BaseBranch` is master), the start index will be 2, and both `C'` and `C` will be included in `commits[:startIdx]`.

**Review by ethantkoenig — CHANGES_REQUESTED** [review]

File: models/issue_comment.go
Comment (ethantkoenig) [reply models/issue_comment.go:676]: Also, it is possible that the head commit of the base branch is not in `commits` at all.

Suppose I have
```
A(master, origin/master)
 \
  B(develop, origin/develop)
```

and I replace commit `B` with a new commit `C` locally
```
  C(develop)
 /
A(master, origin/master)
 \
  B(origin/develop)
```

If I force push develop, `opts.Commits` will contain `[C, B]`, in that order. In this case, `startIdx` will still be -1 after the for loop, and `commits` will still contain both `C` and `B`, which is incorrect.

Comment (lunny) [reply models/issue_comment.go:676]: C will not in the commits list since it has been commit before. See https://github.com/go-gitea/gitea/blob/master/models/update.go#L257

**Review by ethantkoenig — CHANGES_REQUESTED** [review]

File: models/issue_comment.go
Comment (ethantkoenig) [reply models/issue_comment.go:676]: @lunny I have added print statements to the beginning of `CommitRepoActions` (which is run after line 257 of `models/update.go`), to print the contents of `opts.Commits`. In both of the cases described in my comments, I have gotten the results that match what my comments describe. If you are unable to reproduce these results, or if I am missing something, please let me know.

**Comment by techknowlogick** [comment]
@lunny closing as outdated. Please re-open if needed.

**Pull request closed without merging**
